pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java

```diff
@@ -151,30 +144,68 @@ private void createTextIndexForColumn(ColumnMetadata columnMetadata)
     // further need to move around the lucene index directory since it is created with correct directory structure
     // based on segmentVersion.
-    try (LuceneTextIndexCreator textIndexCreator = new LuceneTextIndexCreator(column, segmentDirectory, true);
-        VarByteChunkSVForwardIndexReader forwardIndexReader = getForwardIndexReader(columnMetadata);
-        ChunkReaderContext readerContext = forwardIndexReader.createContext()) {
-      for (int docId = 0; docId < numDocs; docId++) {
-        textIndexCreator.addDoc(forwardIndexReader.getString(docId, readerContext), docId);
+    try (ForwardIndexReader forwardIndexReader = getForwardIndexReader(columnMetadata);
+        ForwardIndexReaderContext readerContext = forwardIndexReader.createContext();
+        LuceneTextIndexCreator textIndexCreator = new LuceneTextIndexCreator(column, segmentDirectory, true)) {
+      if (!hasDictionary) {
+        // text index on raw column, just read the raw forward index
+        VarByteChunkSVForwardIndexReader rawIndexReader = (VarByteChunkSVForwardIndexReader)forwardIndexReader;
+        ChunkReaderContext chunkReaderContext = (ChunkReaderContext)readerContext;
+        for (int docId = 0; docId < numDocs; docId++) {
+          textIndexCreator.addDoc(rawIndexReader.getString(docId, chunkReaderContext), docId);
+        }
+      } else {
+        // text index on dictionary encoded SV column
+        // read forward index to get dictId
+        // read the raw value from dictionary using dictId
+        try (BaseImmutableDictionary dictionary = getDictionaryReader(columnMetadata)) {
+          if (columnMetadata.isSingleValue()) {
+            for (int docId = 0; docId < numDocs; docId++) {
+              int dictId = forwardIndexReader.getDictId(docId, readerContext);
+              String value = dictionary.getStringValue(dictId);
+              textIndexCreator.addDoc(value, docId);
+            }
+          }
+        }
       }
       textIndexCreator.seal();
     }
+
     LOGGER.info("Created text index for column: {} in segment: {}", column, _segmentName);
     PropertiesConfiguration properties = SegmentMetadataImpl.getPropertiesConfiguration(_indexDir);
     properties.setProperty(getKeyFor(column, TEXT_INDEX_TYPE), TextIndexType.LUCENE.name());
     properties.save();
   }
 
-  private VarByteChunkSVForwardIndexReader getForwardIndexReader(ColumnMetadata columnMetadata)
+  private ForwardIndexReader<?> getForwardIndexReader(ColumnMetadata columnMetadata)
+      throws IOException {
+    if (!columnMetadata.hasDictionary()) {
+      // raw index
+      PinotDataBuffer buffer = _segmentWriter.getIndexFor(columnMetadata.getColumnName(), ColumnIndexType.FORWARD_INDEX);
+      return new VarByteChunkSVForwardIndexReader(buffer, DataType.STRING);
+    } else {
+      PinotDataBuffer buffer = _segmentWriter.getIndexFor(columnMetadata.getColumnName(), ColumnIndexType.FORWARD_INDEX);
```

Comment:
Should it be `ColumnIndexType.DICTIONARY`?

Reply:
No